src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c

```diff
@@ -0,0 +1,626 @@
+ *      "etag": "IjJmMDBjNjEzLTAwMDAtNGQwMC0wMDAwLTY1MDhkZDcxMDAwMCI="
+ *   }
+ *
+ * If the registration status is "assigned", then the device is registered
+ *
+ *
+ */
+bool ProcessDeviceRegistrationResponse(const char* payload, size_t payload_len)
+{
+    bool result = false;
+
+    JSON_Value* root_value = json_parse_string(payload);
+    if (root_value == NULL)
+    {
+        Log_Error("Failed to parse JSON payload");
+        goto done;
+    }
+
+    const char* status = json_object_get_string(json_object(root_value), "status");
+    if (status == NULL)
+    {
+        Log_Error("Failed to get status from JSON payload");
+        goto done;
+    }
+
+    if (strncmp(status, "assigned", 8) == 0)
+    {
+        Log_Info("Device is registered.");
+        SetRegisterState(ADPS_REGISTER_STATE_REGISTERED, "received assigned status");
+    }
+    else if (strncmp(status, "assigning", 9) == 0)
+    {
+        Log_Info("Device is registering.");
+        SetRegisterState(ADPS_REGISTER_STATE_REGISTERING, "received assigning status");
+    }
+    else
+    {
+        Log_Error("Unknown status: %s", status);
+        SetRegisterState(ADPS_REGISTER_STATE_FAILED, "received unknown status");
+        // TODO (nox-msft) : determine whether to retry or bail
+    }
+
+done:
+    return result;
+}
+
+/* Callback called when the client receives a message. */
+void ADPS_MQTT_Client_Module_OnMessage(
+    struct mosquitto* mosq, void* obj, const struct mosquitto_message* msg, const mosquitto_property* props)
+{
+    bool result = false;
+    char** topics = NULL;
+    int count = 0;
+    Log_Info("on_message: Topic: %s; QOS: %d; mid: %d", msg->topic, msg->qos, msg->mid);
+
+    if (msg != NULL && msg->topic != NULL)
+    {
+        time_t nowTime = ADUC_GetTimeSinceEpochInSeconds();
+
+        int mqtt_ret = mosquitto_topic_matches_sub(ADPS_MQTT_TOPIC_REGISTRATIONS_RESULT, msg->topic, &result);
+        if (mqtt_ret != MQTT_RC_SUCCESS || !result)
+        {
+            goto done;
+        }
+
+        // Get http status code from topic sub segment
+        // e.g., "$dps/registrations/res/200/?$rid=1&..."
+        mqtt_ret = mosquitto_sub_topic_tokenise(msg->topic, &topics, &count);
+        if (mqtt_ret != MQTT_RC_SUCCESS && count <= 3)
+        {
+            // Cant get the http status code.
+            goto done;
+        }
+
+        int http_status_code = atoi(topics[3]);
+        if (http_status_code == 200)
+        {
+            Log_Info("Device is registered.");
+            s_moduleState.lastRegisterResponseTime = nowTime;
+            ProcessDeviceRegistrationResponse(msg->payload, msg->payloadlen);
+        }
+        else if (http_status_code == 202)
+        {
+            Log_Info("Sending status polling request.");
+            // const char* status = NULL;
+            JSON_Value* root_value = json_parse_string(msg->payload);
+            if (root_value != NULL)
+            {
+                free(s_moduleState.operationId);
```

Comment:
Does operationId have to be another obj on the heap, or could it just copy the string into fixed array set aside for the current one to avoid further fragmenting the heap?